### PR TITLE
OptimizerAPR permissionless manualAllocation

### PR DESCRIPTION
**Pull request by GuillaumeNervoXS**

### CODE REVIEW

File: contracts/strategies/OptimizerAPR/genericLender/euler/GenericEulerStaker.sol [hunk 1]
```diff
@@ -19,6 +19,9 @@ contract GenericEulerStaker is GenericEuler, OracleMath {
     uint256 internal constant _SECONDS_IN_YEAR = 365 days;
     uint32 internal constant _TWAP_PERIOD = 1 minutes;
 
+    /// @notice EUL token address
+    IERC20 private constant _EUL = IERC20(0xd9Fcd98c322942075A5C3860693e9f4f03AAE07b);
```
Comment (sogipec) [inline contracts/strategies/OptimizerAPR/genericLender/euler/GenericEulerStaker.sol:22]: Where are we using it in the contract?

Comment (GuillaumeNervoXS) [reply contracts/strategies/OptimizerAPR/genericLender/euler/GenericEulerStaker.sol:22]: line 52

File: contracts/strategies/BaseStrategyUpgradeable.sol [hunk 2]
```diff
@@ -50,14 +50,15 @@ abstract contract BaseStrategyUpgradeable is BaseStrategyEvents, AccessControlAn
     uint256 public constant BASE = 10**18;
     uint256 public constant SECONDSPERYEAR = 31556952;
```
Comment (greedythib) [inline contracts/strategies/BaseStrategyUpgradeable.sol]: @GuillaumeNervoXS why not using underscores?

File: contracts/strategies/BaseStrategyUpgradeable.sol [hunk 3]
```diff
@@ -111,7 +112,8 @@ abstract contract BaseStrategyUpgradeable is BaseStrategyEvents, AccessControlAn
         _setRoleAdmin(GUARDIAN_ROLE, POOLMANAGER_ROLE);
 
         // Initializing roles first
-        for (uint256 i = 0; i < keepers.length; i++) {
+        uint256 keepersLength = keepers.length;
```
Comment (greedythib) [inline contracts/strategies/BaseStrategyUpgradeable.sol:110]: @GuillaumeNervoXS is it cost efficient to store the length like this?

Comment (Picodes) [reply contracts/strategies/BaseStrategyUpgradeable.sol:110]: it is, you avoid rereading it in the for loop

File: contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol [hunk 4]
```diff
@@ -0,0 +1,491 @@
...
+    function manualAllocation(LenderRatio[] memory _newPositions) external onlyRole(GUARDIAN_ROLE) {
+        IGenericLender[] memory lendersList = lenders;
+        uint256 share = 0;
+        for (uint256 i = 0; i < lendersList.length; i++) {
+            lendersList[i].withdrawAll();
+        }
+
+        uint256 assets = want.balanceOf(address(this));
+
+        for (uint256 i = 0; i < _newPositions.length; i++) {
```
Comment (greedythib) [inline contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol:378]: `++i`

File: contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol [hunk 5]
```diff
@@ -0,0 +1,491 @@
...
+            bool found = false;
+
+            //might be annoying and expensive to do this second loop but worth it for safety
+            for (uint256 j = 0; j < lendersList.length; j++) {
```
Comment (greedythib) [inline contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol:382]: ++j

File: contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol [hunk 6]
```diff
@@ -0,0 +1,491 @@
...
+    }
+
+    /// @notice Reallocates all funds according to a new distributions
+    /// @param _newPositions List of shares to specify the new allocation
+    /// @dev Share must add up to 1000. 500 means 50% etc
+    /// @dev This code has been forked, so we have not thoroughly tested it on our own
+    function manualAllocation(LenderRatio[] memory _newPositions) external onlyRole(GUARDIAN_ROLE) {
+        IGenericLender[] memory lendersList = lenders;
+        uint256 share = 0;
+        for (uint256 i = 0; i < lendersList.length; i++) {
```
Comment (greedythib) [inline contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol:372]: ++i

File: contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol [hunk 7]
```diff
@@ -336,72 +385,63 @@ contract OptimizerAPRStrategy is BaseStrategyUpgradeable {
...
             return 0;
         }
 
-        uint256 weightedAPR = 0;
-
-        for (uint256 i = 0; i < lenders.length; i++) {
+        uint256 weightedAPR;
+        uint256 lendersLength = lenders.length;
+        for (uint256 i; i < lendersLength; ++i) {
             weightedAPR = weightedAPR + lenders[i].weightedApr();
```
Comment (greedythib) [inline contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol]: why not `weightedAPR += lenders[i].weightedApr()`?

File: contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol [hunk 8]
```diff
@@ -131,95 +134,147 @@ contract OptimizerAPRStrategy is BaseStrategyUpgradeable {
...
+        if (_totalApr < estimatedAprHint) {
+            for (uint256 i; i < lendersListLength; ++i) {
+                if (lenderAdjustedAmounts[i] < 0) lendersList[i].withdraw(uint256(-lenderAdjustedAmounts[i]));
+            }
 
-        uint256 bal = want.balanceOf(address(this));
-        if (bal > 0) {
-            want.safeTransfer(address(lendersList[highest]), bal);
-            lendersList[highest].deposit();
+            for (uint256 i; i < lendersListLength; ++i) {
```
Comment (greedythib) [inline contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol:254]: I am not sure to understand why we cannot combine both for loops together

File: contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol [hunk 9]
```diff
@@ -131,95 +134,147 @@ contract OptimizerAPRStrategy is BaseStrategyUpgradeable {
     /// @notice Estimates highest and lowest apr lenders among a `lendersList`
     /// @param lendersList List of all the lender contracts associated to this strategy
     /// @return _lowest The index of the lender in the `lendersList` with lowest apr
-    /// @return _lowestApr The lowest apr
     /// @return _highest The index of the lender with highest apr
-    /// @return _potential The potential apr of this lender if funds are moved from lowest to highest
+    /// @return _investmentStrategy Whether we should invest from the lowest to the highest yielding strategy or simply invest loose assets
     /// @dev `lendersList` is kept as a parameter to avoid multiplying reads in storage to the `lenders`
     /// array
```
Comment (greedythib) [inline contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol]: `/// @return _totalApr The APR computed according to (greedy) heuristics that will determine whether positions are adjusted or not`

File: contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol [hunk 10]
```diff
@@ -249,7 +304,7 @@ contract OptimizerAPRStrategy is BaseStrategyUpgradeable {
             uint256 amountWithdrawnFromStrat = lendersList[lowest].withdraw(_amount - amountWithdrawn);
             // To avoid staying on the same strat if we can't withdraw anythin from it
             amountWithdrawn = amountWithdrawn + amountWithdrawnFromStrat;
```
Comment (greedythib) [inline contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol]: why not `amountWithdrawn += amountWithdrawnFromStrat`

File: contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol [hunk 11]
```diff
@@ -0,0 +1,491 @@
...
+
+    /// @notice Estimates highest and lowest apr lenders among a `lendersList`
+    /// @param lendersList List of all the lender contracts associated to this strategy
+    /// @return _lowest The index of the lender in the `lendersList` with lowest apr
+    /// @return _lowestApr The lowest apr
+    /// @return _highest The index of the lender with highest apr
+    /// @return _potential The potential apr of this lender if funds are moved from lowest to highest
+    /// @dev `lendersList` is kept as a parameter to avoid multiplying reads in storage to the `lenders`
+    /// array
+    function _estimateAdjustPosition(IGenericLender[] memory lendersList)
```
Comment (greedythib) [inline contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol:142]: the naming here is a little bit confusing and I think we should call it the same way as  in `OptimizerAPRStrategy.sol` as both heuristics are greedy. Maybe there was a reason for it?

Comment (sogipec) [reply contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol:142]: In the `OptimizerAPRGreedyStrategy` file we just copy pasted the old state of the code as a way to keep it and track through the tests to what extent we were improving the baseline.
That's why we haven't done the gas optimization there

Comment (greedythib) [reply contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol:142]: ok so this contract is just an old version that should not be meant for prod?

Comment (sogipec) [reply contracts/strategies/OptimizerAPR/OptimizerAPRGreedyStrategy.sol:142]: Yes correct

File: contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol [hunk 12]
```diff
@@ -131,110 +137,175 @@ contract OptimizerAPRStrategy is BaseStrategyUpgradeable {
...
+            for (uint256 i; i < lendersListLength; ++i) {
+                if (lenderAdjustedAmounts[i] < 0) {
+                    deltaWithdraw +=
+                        uint256(-lenderAdjustedAmounts[i]) -
+                        lendersList[i].withdraw(uint256(-lenderAdjustedAmounts[i]));
+                }
+            }
 
-        (uint256 lowest, uint256 lowestApr, uint256 highest, uint256 potential) = _estimateAdjustPosition(lendersList);
+            // Strategy didn't succeed to withdraw the intended funds --> go through the greedy path
```
Comment (sogipec) [inline contracts/strategies/OptimizerAPR/OptimizerAPRStrategy.sol]: If the strategy didn't succeed to withdraw the intended funds -> revert and force the greedy path

**Review by sogipec — APPROVED** [review]

**Pull request merged**